ARCs/arc-0084.md

```diff
@@ -0,0 +1,750 @@
+
+### Separation of Transfer and Data Apps: Data Access In Transfer Hooks
+
+In order to accomplish the above point of trusted accounting logic and flexibility, we need to have an external app, the **Transfer Hook App**. This app can approve or deny transfers for a given token and app may also execute additional logic like snapshots or royalty enforcement. In order to do these functions effectively, it must have access to the token's data, which is held in the **Data App**. If the data app implement the transfer method, and thus the call to the transfer hook app directly, reading token data wouldn't be possible because re-entrancy is not allowed in the AVM. Having two apps lets us have one app, the transfer app, be responsible for transfers and calling the transfer hook app, while the data app is solely responsible for the accounting and storing data. This gives us full programmability without any re-entrancy.  
+
+### Multiple Transfers Per Call: Atomicity & Readability
+
+Allowing multiple transfers in one method calls allows for more efficient transfers. A single method call can transfer any number of tokens between any number of accounts (within the resource limitations of the AVM). More importantly, having multiple transfers in a single calls allows for inspection of transfer groups by the **Transfer Hook App** that would otherwise be impossible due to the fact that inner transactions cannot access high-level transaction groups. This enables possibilities such as denying any transfer that doesn't include a royalty payment.
+
+### Allowances With Expiration & Cooldown: Safely Enable Delegation Features
+
+Approvals are a key feature of other smart contract tokens, such as ERC20 and ARC-200, but they have serious security implications. For example, a large amount of EVM exploits occur due to users giving protocols token approvals and that same protocol being exploited years later after they no longer us the protocol. This is the driving rationale for having an expiration time for allowances in ARC-84. By setting an expiration time, users can be sure a protocol will not have access to their tokens when they are done using the app without having to take any explicit action. Similarly, the cooldown mechanism enables a highly-requested feature, subscriptions, in a secure manner. By being supported directly in ARC-84, there is no implementation ambiguity and all apps in the ecosystem can clearly display allowance information to users.
```

Comment:
AFAIU allowances and approvals exist in ERC20 mostly to work around the fact that there are no transaction groups in the EVM. It is not possible for example to do a swap by doing a transfer of USDC + swap call in one transaction. Instead the swap contract is given approval to spend USDC, followed separately by a contract call to swap usdc to something. 

As I mentioned above, subscriptions and other allowance based features seem more appropriate for a wallet app standard than an asset standard.